lib/engine/round/g_1824/stock.rb

```diff
@@ -37,6 +37,58 @@ def next_entity_index!
 
           @entity_index = (@entity_index - 1) % @entities.size
         end
+
+        def finish_round
+          @game.log << 'First stock round is finished - any unsold Pre-State Railways, Coal Railways, ' \
+            ' and Montain Railways are removed from the game'
+
+          @game.purchasable_unsold_companies.select { |c| @game.mountain_railway?(c) }.each do |m|
+            @game.log << "Mountain Railway #{m.name} closes"
+            m.close!
+          end
+
+          @game.purchasable_unsold_companies.each do |p|
+            pre_state = @game.minor_by_id(p.id)
+            state = @game.associated_state_railway(p)
+            @game.log << "Pre-Staatsbahn Railway #{pre_state.name} closes; "\
+              "unreserved corresponding share in #{state.name} is no longer reserved"
+
+            # Remove home city reservation
+            remove_reservation(pre_state)
+
+            # Remove home token
+            pre_state.tokens.first.remove!
+
+            pre_state.close!
+            p.close!
+          end
+
+          @game.corporations.select { |c| @game.coal_railway?(c) }.reject(&:floated?).each do |c|
```

Comment:
just do a next if, you're doing 3 passes over the data here

Reply:
Have refactored a bit so it is one loop over companies and one loop over corporations.